demos/twa-offline-first/src/main/java/com/google/androidbrowserhelper/demos/twa_offline_first/OfflineFirstTWALauncherActivity.java

```diff
@@ -38,27 +40,47 @@ protected boolean shouldLaunchImmediately() {
     }
 
     private void tryLaunchTwa() {
-        // Check connection status. If online, launch the Trusted Web Activity with `launchTwa()`.
-        // if offline render the offline fallback screen.
-        if (isOnline()) {
+        // If TWA has already launched successfully, launch TWA immediately.
+        // Otherwise, check connection status. If online, launch the Trusted Web Activity with `launchTwa()`.
+        // Otherwise, if offline, render the offline fallback screen.
+        if (hasTwaLaunchedSuccessfully()) {
             launchTwa();
+        } else if (isOnline()) {
+            firstTimeLaunchTwa();
         } else {
             renderOfflineFallback();
         }
     }
 
+    private boolean hasTwaLaunchedSuccessfully() {
+        // Return true if the preference "twa_launched_successfully" has already been set.
+        SharedPreferences sharedPref = getSharedPreferences(getString(R.string.twa_offline_first_preferences_file_key), Context.MODE_PRIVATE);
```

Comment:
This may cause a StrictMode violation. Generally the first time you access SharedPreferences it does a disk read, but then for subsequent times it's fine. There's not really much we can do about it though - reading that value from the file is on our critical path, so we need to do it.

In order to stop the demo app from complaining about this on people's phones, we may need to surround this line with:

```
StrictMode.ThreadPolicy oldPolicy = StrictMode.getThreadPolicy();
try {
  StrictMode.setThreadPolicy(StrictMode.ThreadPolicy.LAX);
  
  // Do the disk read here.
} finally {
  StrictMode.setThreadPolicy(oldPolicy);
}
```